### DISCUSSION TITLE
How to create a repository with a pre-defined filter via a DI'd service (Nest)?

### DISCUSSION THREAD

**Discussion by Seqi** [mikro-orm/mikro-orm#2445] [Q&A] [answered]
Using nest.js and mikroorm, I have around 9 entities which all follow the same base entity, something that contains `createdAt`, `id` and `uid`. Every single one of these entities want to have `filterBy: uid = current authenticated user` and `orderBy: createdAt DESC`. To prevent having to copy paste code everywhere, I wish to make a generic service i can either inject, extend from, etc. 

My base entity is as follows:
```
export abstract class UserBaseEntity extends BaseEntity {
	@PrimaryKey({ length: 14 })
	id: string = nanoid(14)

	@Property({ fieldName: 'createdAt', length: 6 })
	createdAt!: Date

	@Property({ length: 32 })
	uid!: string
}
```

The approaches i've tried so far are:

**Create a generic-driven service**

```
@Injectable()
export class UserResourceService<TResource extends UserBaseEntity> {
	constructor(
		private repo: EntityRepository<TResource>,
		private user: UserService,
	) {}

	async getAll(): Promise<TResource[]> {
		const result = await this.repo.find(
			{
				uid: this.user.uid, // Argument of type '{ uid: string; }' is not assignable to parameter of type 'FilterQuery<TResource>'.
			},
			{ orderBy: { createdAt: QueryOrder.ASC } },
		)

		return result
	}
```

This fails at the filter line, as it does not see `uid` on `TResource`. 

**Create a custom repository**

I tried to extend from repository like so:

```
export class ResourceService<
	TResource extends UserBaseEntity,
> extends EntityRepository<TResource> {
	constructor(
		private repo: EntityRepository<TResource>,
		private user: UserService,
	) {
              // how?
             super(...
	}
```

This fails as I need to call super with several arguments that DI container would usually handle. I could figure these out but it seems like not the right way to do this.

**Use the new filter decorator**
Finally, I have tried using the filter decorator like so:

```

@Filter<UserBaseEntity>({ name: 'belongsTo', cond: { uid: // Inject user sevice somehow? }  })
export abstract class UserBaseEntity extends BaseEntity {
```
This hasn't worked for me so far as I haven't been able to get the DI container in this decorator. 

For something that seems relatively simple - always filtering a specific base entity by a user - what is the best way to achieve this using mikro orm and nest js?

**Comment by B4nan** [comment, marked as answer]
You can use global filters:

https://mikro-orm.io/docs/filters/#global-filters

> Argument of type '{ uid: string; }' is not assignable to parameter of type 'FilterQuery<TResource>'.

This is a TS problem, you can get around it only via type casting. Nothing we can do about.

**Reply by Seqi** [reply]
Thanks for the super fast reply. Type cast could be a good idea, completely slipped my mind.

I like the global filter idea though - I've read through the docs and it seems to be either all or explicit. If I pass in an abstract entity, will mikro-orm know to apply it to all entities that extend that base entity? Or perhaps some logic in a target-all filter?

**Reply by B4nan** [reply]
there are no base entities on runtime, their metadata are inlined to the children. so filters on base entities are the same as if you would define them on every extending entity.

**Reply by Seqi** [reply]
So global filters seems to be the go-to for this. I tried using the base entity name, but the filter doesnt seem to apply. The code i'm using is:

```
	const em = app.get<EntityManager>(EntityManager)
	em.addFilter('test', { uid: '123' }, [UserBaseEntity])
```

123 does not match any user id. When performing a `count`, I get 36 total. When I swap this to:

```
	const em = app.get<EntityManager>(EntityManager)
	em.addFilter('test', { uid: '123' }, [Concrete])
```

It correctly gets 0.

If it makes a difference, there are several levels of subclasses.

```
class Concrete extends Base1 { }
class Base1 extends UserBaseEntity
```

Did I miss a step?

**Reply by B4nan** [reply]
As I already told you, there are no base entities on runtime, so you cant use your base entity to define a filter on runtime. Just define the filter via decorator on the base entity, and enable it by default. Or list the entities that are extending it explicitly, but that could be error prone (you could use the ORM metadata, but still).

```ts
@Filter({ name: '...', cond: args => ..., default: true })
export class UserBaseEntity { }
```

**Reply by Seqi** [reply]
Ah sorry I misunderstood. You're right, I'm trying to avoid manually setting them all. I've done as you suggested and that seems to work well.

As a follow up question - what is the scope of the parameter being set? As the docs suggest, I plan on setting this parameter via middleware to add the authenticated user as the param. However - once this request is done, does the parameter 'clear' so to speak? Or will further requests that don't set this parameter still persist with the previously set one?

**Reply by B4nan** [reply]
You will set it on the EM fork that exists only for given context (request).

**Reply by Seqi** [reply]
Makes sense. Thanks a lot for all your help.
